### DISCUSSION TITLE
Migrating from Prisma+Nexus to Prisma+Pothos

### DISCUSSION THREAD

**Discussion by baconcheese113** [hayes/pothos#511] [Q&A] [answered]
Hey, thanks for all your work on pothos!
I've been trying to migrate to Pothos for a couple days and reading through all the docs but I'm a bit stuck in a few places.

## Unable to change context shape

All Query entrypoints require an Auth token, so all `prismaObject`s should be using my `IAuthContext` instead of the `IContext`. Before I would just cast the context to `IAuthContext` in the resolvers, but it seems like there's a better way if I use `ScopeAuth`. So I tried this:
```ts
export const builder = new SchemaBuilder<{
    Context: IContext;
    AuthScopes: { loggedIn: boolean }
    AuthContexts: { loggedIn: IAuthContext }
    ...
}>({
    plugins: [ScopeAuth, PrismaPlugin],
    ...
    scopeAuthOptions: {
        runScopesOnType: true,
    },
    authScopes: (context) => ({
        loggedIn: !!context.user || !!context.hub
    })
})
```
along with copy pasting this across all my prismaObjects
```ts
builder.prismaObject('Hub', {
    authScopes: {loggedIn: true},
    runScopesOnType: true,
    fields: (t) => ({
    ...
})
```
But I can't get the context to change from `IContext`

## Exposing ID

Just a question about `exposeID`, will it expose an `Int!` type if that's what `id` is defined as in the schema?

## Viewer field (non-prisma models)

One of my `Query` entrypoints is Query->viewer->user->etc... like this:
```ts
builder.queryType({
    fields: (t) => ({
        ....
        viewer: t.field({
            type: Viewer,
            resolve(_root, _args, { user }) {
                if (!user) throw new AuthenticationError("User does not have permission")
                
                return ({}) // <--------------is this correct?
            } 
        }),
    })
})

export class Viewer {
    constructor(
        user: User,
        hubs: Hub[],
        networks: Network[],
        latestSensorVersion: string,
    ) {}
}
builder.objectType(Viewer, {
    fields: (t) => ({
        user: t.prismaField({
            type: "User",
            resolve: (query, _root, _args, { user }) => {
                return user
            },
        }),
        ...
    })
})
```
but I get this error for the viewer object `fields`
```
Argument of type '{ fields: (t: ObjectFieldBuilder<ExtendDefaultTypes<{ Context: IContext; AuthScopes: { loggedIn: boolean; }; AuthContexts: { loggedIn: IAuthContext; }; PrismaTypes: PrismaTypes; Scalars: { ...; }; }>, Viewer>) => { ...; }; }' is not assignable to parameter of type 'ObjectTypeOptions<ExtendDefaultTypes<{ Context: IContext; AuthScopes: { loggedIn: boolean; }; AuthContexts: { loggedIn: IAuthContext; }; PrismaTypes: PrismaTypes; Scalars: { DateTime: { ...; }; }; }>, typeof Viewer, Viewer, InterfaceParam<...>[]>'.
  Type '{ fields: (t: ObjectFieldBuilder<ExtendDefaultTypes<{ Context: IContext; AuthScopes: { loggedIn: boolean; }; AuthContexts: { loggedIn: IAuthContext; }; PrismaTypes: PrismaTypes; Scalars: { ...; }; }>, Viewer>) => { ...; }; }' is missing the following properties from type '{ name: string; interfaces: InterfaceParam<ExtendDefaultTypes<{ Context: IContext; AuthScopes: { loggedIn: boolean; }; AuthContexts: { loggedIn: IAuthContext; }; PrismaTypes: PrismaTypes; Scalars: { ...; }; }>>[] | (() => InterfaceParam<...>[]); ... 7 more ...; skipInterfaceScopes?: boolean | undefined; }': name, interfaces
```

**Comment by hayes** [comment, marked as answer]
> All Query entrypoints require an Auth token, so all prismaObjects should be using my IAuthContext instead of the IContext. Before I would just cast the context to IAuthContext in the resolvers, but it seems like there's a better way if I use ScopeAuth. So I tried this:

If all entrypoints guarantee a user, you can use just make your auth context the default context:

```ts
export const builder = new SchemaBuilder<{
    Context: IAuthContext;
    AuthScopes: { loggedIn: boolean }
}>({...})
```

The plugin isn't currently changing the context type automatically for fields of a type that have defined scopes.  To change the context based on scopes you need to use either the `t.authField` method, or the `t.withAuth` method so that the field definitions become aware of scopes being used and can automatically type the context based on the scopes. 

also, just a note: if you have `runScopesOnType: true,` set in the builder option, you don't need to repeat it on every type.


> Just a question about exposeID, will it expose an Int! type if that's what id is defined as in the schema?

No, this will be exoised as in `ID` field (nullability can be changed, defaults to nonnullable).  You can use t.exposeInt if you want to create an `Int` field in your schema. 

> One of my Query entrypoints is Query->viewer->user->etc... like this:

When you use a class for an object type like: `builder.objectType(Viewer, {` fields that return that type (like the viewer query) need to return an instance of that class.  In this case, you resolver should return something like `new Viewer((user, hubs, networks, latestSensorVersion)`

**Reply by hayes** [reply]
alternatively, it it looks like your Viewer class isn't doing anything, so you can instead just not use a class and continue returning en empty object: 
```ts
builder.objectRef<{} /* this is the type you need to return */>('Viewer').implement({
  fields: t => ({
    user: t.prismaField({...})
  }),
})
`

**Reply by baconcheese113** [reply]
> No, this will be exoised as in ID field

Ah, that clears that up!

---

> so you can instead just not use a class and continue returning an empty object

That fixed the errors on the object, then for the viewer field `type` I wasn't able to use `{ }` or `'Viewer'` as the type, but it worked by referencing the return value of `builder.objectRef` like this:
```ts
const Viewer = builder.objectRef<{}>('Viewer').implement({
  fields: t => ({
    user: t.prismaField({...})
  }),
})

builder.queryType({
    fields: (t) => ({
        ....
        viewer: t.field({
            type: Viewer,
            resolve: (_root, _args, { user }) => {
                if (!user) throw new AuthenticationError("User does not have permission")
                return ({})
            } 
        }),
    })
})
```

---

> To change the context based on scopes you need to use either the t.authField method, or the t.withAuth method so that the field definitions become aware of scopes being used and can automatically type the context based on the scopes.

Oh I see, this could get a bit verbose. Currently my `queryType` fields and the majority of my mutations check for auth in the resolver and throw if `user` is undefined. If I were to change the default Context, then I guess I'd need to set it individually on the mutations and `queryType` fields instead?

Is there a better way to do what I'm trying to do? If not, are there any plans to add a way to configure the context shape at the type or object level?

**Reply by hayes** [reply]
No current plans.  I think you could probably cast the builder.  Eg `const authedBuilder = builder as unknown as PothosSchemaTypes.SchemaBuilder<PothosSchemaTypes.ExtendDefaultTypes<Types>>`. With Types being the same types passed to the original builder (and replacing context).  This is more of a workaround, but should let you define objects with that version of the builder without having to cast the context type in every resolver. 

Hope that helps

**Reply by hayes** [reply]
Would be nice to change the context on types automatically, but implementing it with the current plugin API is challenging.  For example, it would be easy to add a builder.authedObject method, but then you would also need a builder.authedPrismaObject (and a whole lot of other ones).

**Reply by baconcheese113** [reply]
Got all those resolved but hit an issue trying to finally start the server or generate types
```
C:\Users\bacon\Documents\WebDev\Personal\handle-it\server\node_modules\@pothos\core\src\config-store.ts:231
      throw new Error(`Duplicate typename: Another type with name ${name} already exists.`);
Error: Duplicate typename: Another type with name Mutation already exists.
    at ConfigStore.addTypeConfig (C:\Users\bacon\Documents\WebDev\Personal\handle-it\server\node_modules\@pothos\core\src\config-store.ts:231:13)
    at SchemaBuilder.mutationType (C:\Users\bacon\Documents\WebDev\Personal\handle-it\server\node_modules\@pothos\core\src\builder.ts:247:22)
    at Object.<anonymous> (C:\Users\bacon\Documents\WebDev\Personal\handle-it\server\src\resolvers\mutations\hubMutations.ts:4:9)
    at Module.m._compile (C:\Users\bacon\Documents\WebDev\Personal\handle-it\server\node_modules\ts-node\src\index.ts:1597:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1157:10)
    at Object.require.extensions.<computed> [as .ts] (C:\Users\bacon\Documents\WebDev\Personal\handle-it\server\node_modules\ts-node\src\index.ts:1600:12)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:13)
```
My mutations are broken up into separate files that define all the mutations related to a type, so a typical definition is like:
```ts
// hubMutations.ts
builder.mutationType({
    fields: (t) => ({
        createHub: t.prismaField({
            ....
        }),
        deleteHub: t.prismaField({
            ....
        }),
        updateHub: t.prismaField({
            ....
        }),
        updateHubBatteryLevel: t.prismaField({
            ....
        }),
    })
})
```
Can you see what I'm doing wrong here?

**Reply by hayes** [reply]
You can only define t.mutationType once.  To add more mutation fields, use t.mutationFields instead

**Reply by baconcheese113** [reply]
Oh perfect, that makes sense

Switched all my mutations to something like this:
```ts
// hubMutations.ts
builder.mutationFields((t) => ({
    createHub: t.prismaField({
        ....
    }),
    deleteHub: t.prismaField({
        ....
    }),
    updateHub: t.prismaField({
        ....
    }),
    updateHubBatteryLevel: t.prismaField({
        ....
    }),
}))
```
butI'm getting a different error now for all uses of `withAuth`:
```
C:\Users\bacon\Documents\WebDev\Personal\handle-it\server\src\resolvers\graphTypes\hubGraphType.ts:30
        events: t.withAuth({ loggedIn: true }).prismaField({
                                              ^
TypeError: Cannot read properties of undefined (reading 'prismaField')
    at fields (C:\Users\bacon\Documents\WebDev\Personal\handle-it\server\src\resolvers\graphTypes\hubGraphType.ts:30:47)
    at Object.fields (C:\Users\bacon\Documents\WebDev\Personal\handle-it\server\node_modules\@pothos\plugin-prisma\src\schema-builder.ts:48:11)
    at fields (C:\Users\bacon\Documents\WebDev\Personal\handle-it\server\node_modules\@pothos\core\src\builder.ts:173:24)
    at cb (C:\Users\bacon\Documents\WebDev\Personal\handle-it\server\node_modules\@pothos\core\src\config-store.ts:426:66)
    at ConfigStore.onTypeConfig (C:\Users\bacon\Documents\WebDev\Personal\handle-it\server\node_modules\@pothos\core\src\config-store.ts:350:7)
    at ConfigStore.addFields (C:\Users\bacon\Documents\WebDev\Personal\handle-it\server\node_modules\@pothos\core\src\config-store.ts:425:12)
    at fn (C:\Users\bacon\Documents\WebDev\Personal\handle-it\server\node_modules\@pothos\core\src\config-store.ts:423:45)
    at forEach (C:\Users\bacon\Documents\WebDev\Personal\handle-it\server\node_modules\@pothos\core\src\config-store.ts:401:30)
    at Array.forEach (<anonymous>)
    at ConfigStore.prepareForBuild (C:\Users\bacon\Documents\WebDev\Personal\handle-it\server\node_modules\@pothos\core\src\config-store.ts:401:9)
```

**Reply by hayes** [reply]
That's a weird error, if there is any more info you can provide that might be helpful.  From the stack trace it looks like you are defining a prismaObject, and calling t.withAuth, which exists, but returns undefined.  As far as I can tell, it should be returning the builder.

**Reply by hayes** [reply]
Actually, this looks like a bug.  I just reproduced it, looking into it now

**Reply by hayes** [reply]
Just pushed a fix, should be published in a few minutes

**Reply by baconcheese113** [reply]
Thanks for fixing that so quickly! After upgrading to 
```
"@pothos/plugin-prisma": "^3.21.2",
"@pothos/plugin-scope-auth": "^3.11.2",
```
I'm now able to start the server without errors with everything migrated from Nexus 🙌

**Reply by hayes** [reply]
nice!
